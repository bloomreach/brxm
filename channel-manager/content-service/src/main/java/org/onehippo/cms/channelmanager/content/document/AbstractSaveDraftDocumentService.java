/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.onehippo.cms.channelmanager.content.document;

import java.io.Serializable;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;

import org.onehippo.cms.channelmanager.content.UserContext;
import org.onehippo.cms.channelmanager.content.document.model.Document;
import org.onehippo.cms.channelmanager.content.document.model.DocumentInfo;
import org.onehippo.cms.channelmanager.content.document.util.EditingUtils;
import org.onehippo.cms.channelmanager.content.documenttype.model.DocumentType;
import org.onehippo.cms.channelmanager.content.error.ErrorInfo;
import org.onehippo.cms.channelmanager.content.error.ErrorWithPayloadException;
import org.onehippo.cms.channelmanager.content.error.ForbiddenException;
import org.onehippo.cms.channelmanager.content.error.InternalServerErrorException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.onehippo.cms.channelmanager.content.document.util.EditingUtils.HINT_PUBLISH;
import static org.onehippo.cms.channelmanager.content.document.util.EditingUtils.HINT_REQUEST_PUBLICATION;
import static org.onehippo.cms.channelmanager.content.document.util.EditingUtils.isHintActionTrue;

public abstract class AbstractSaveDraftDocumentService implements SaveDraftDocumentService {

    private static final Logger log = LoggerFactory.getLogger(AbstractSaveDraftDocumentService.class);

    @Override
    public final Document editDraft(final String identifier, final UserContext userContext) {
        validateEditDraft(identifier, userContext);
        return getDraft(identifier, userContext);
    }

    @Override
    public final boolean canEditDraft(final String identifier, final UserContext userContext) {
        return canEditDraft(getHints(identifier, userContext));
    }

    public final DocumentInfo addDocumentInfo(final String identifier, final UserContext userContext, final Document document) {
        final Map<String, Serializable> hints = getHints(identifier, userContext);
        final DocumentInfo documentInfo = document.getInfo();
        documentInfo.setDirty(isDocumentDirty(identifier, userContext, document));
        documentInfo.setCanPublish(isHintActionTrue(hints, HINT_PUBLISH));
        documentInfo.setCanRequestPublication(isHintActionTrue(hints, HINT_REQUEST_PUBLICATION));
        return documentInfo;
    }

    public final DocumentType getDocumentType(final String identifier, final UserContext userContext) {
        final String id = getVariantNodeType(identifier, userContext).orElseThrow(
                () -> new InternalServerErrorException(new ErrorInfo(ErrorInfo.Reason.DOES_NOT_EXIST)));

        try {
            return getDocumentTypeByNodeTypeIdentifier(userContext, id);
        } catch (ErrorWithPayloadException e) {
            log.warn("Failed to retrieve type of document '{identifier: {} }'", identifier, e);
            throw new InternalServerErrorException(new ErrorInfo(ErrorInfo.Reason.SERVER_ERROR));
        }
    }

    public final boolean canEditDraft(final Map<String, Serializable> hints) {
        return EditingUtils.isHintActionTrue(hints == null ? Collections.emptyMap() : hints, "editDraft");
    }

    abstract boolean isDocumentDirty(String identifier, UserContext userContext, Document document);

    abstract ErrorInfo withDisplayName(ErrorInfo errorInfo, String identifier, UserContext context);

    abstract ErrorInfo withDocumentInfo(ErrorInfo errorInfo, String identifier, UserContext userContext);

    abstract Optional<String> getVariantNodeType(String identifier, UserContext userContext);

    abstract DocumentType getDocumentTypeByNodeTypeIdentifier(UserContext context, String nodeTypeIdentifier);

    abstract Optional<ErrorInfo> determineEditingFailure(Map<String, Serializable> hints, UserContext userContext);

    abstract Map<String, Serializable> getHints(String identifier, UserContext userContext);

    private void validateEditDraft(final String identifier, final UserContext userContext) {
        final Map<String, Serializable> hints = getHints(identifier, userContext);
        if (!canEditDraft(hints)) {
            throw determineEditingFailure(hints, userContext)
                    .map(errorInfo -> withDocumentInfo(errorInfo, identifier, userContext))
                    .map(ForbiddenException::new)
                    .orElseGet(() -> new ForbiddenException(new ErrorInfo(ErrorInfo.Reason.SERVER_ERROR)));
        }

        final DocumentType docType = getDocumentType(identifier, userContext);
        if (docType.isReadOnlyDueToUnsupportedValidator()) {
            final ErrorInfo errorInfo = new ErrorInfo(ErrorInfo.Reason.CREATE_WITH_UNSUPPORTED_VALIDATOR);
            throw new ForbiddenException(withDisplayName(errorInfo, identifier, userContext));
        }
    }

    abstract Document getDraft(String identifier, UserContext userContext);
}
