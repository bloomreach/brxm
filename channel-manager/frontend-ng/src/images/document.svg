<?xml version="1.0" encoding="utf-8"?>
<!--
  Copyright 2016-2018 Hippo B.V. (http://www.onehippo.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px"
     viewBox="0 0 24 24" xml:space="preserve" fill="currentcolor">
  <path clip-rule="evenodd" d="M 18.119 23 L 5.881 23 C 4.842 23 4 22.143 4 21.088 L 4 14.895 L 4 2.912 C 4 1.856 4.842 1 5.881 1 C 5.882 1 14.813 1 14.813 1 L 14.824 1 C 14.974 1 15.103 1.077 15.189 1.189 L 19.815 5.891 C 19.926 5.978 20 6.108 20 6.262 C 20 6.263 20 6.274 20 6.274 L 20 21.088 C 20 22.143 19.158 23 18.119 23 Z M 19.059 6.738 L 14.824 6.738 C 14.563 6.738 14.353 6.524 14.353 6.262 L 14.35 2 L 5.88 2 C 5.36 2 5 2.39 5 2.91 L 5 21.09 C 5 21.61 5.363 22.043 5.88 22 L 18.12 22 C 18.637 22.043 19.059 21.614 19 21.09 L 19 6.74 L 19.059 6.738 Z M 15.293 2.616 L 15.293 5.783 L 18.408 5.783 L 15.293 2.616 Z"/>
  <path d="M 7.679 10 C 7.317 10 7 10.234 7 10.5 C 7 10.767 7.317 11 7.679 11 L 15.821 11 C 16.183 11 16.5 10.767 16.5 10.5 C 16.5 10.234 16.183 10 15.821 10 L 7.679 10 Z"/>
  <path d="M 7.679 14 C 7.317 14 7 14.234 7 14.5 C 7 14.767 7.317 15 7.679 15 L 15.821 15 C 16.183 15 16.5 14.767 16.5 14.5 C 16.5 14.234 16.183 14 15.821 14 L 7.679 14 Z"/>
  <path d="M 7.679 18 C 7.317 18 7 18.234 7 18.5 C 7 18.767 7.317 19 7.679 19 L 15.821 19 C 16.183 19 16.5 18.767 16.5 18.5 C 16.5 18.234 16.183 18 15.821 18 L 7.679 18 Z"/>
</svg>
