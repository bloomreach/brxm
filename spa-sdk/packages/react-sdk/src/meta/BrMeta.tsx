/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaCollection } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: MetaCollection;
}

export class BrMeta extends React.Component<BrMetaProps> {
  private headRef = React.createRef<HTMLElement>();

  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.renderMeta();
  }

  componentDidUpdate(prevProps: BrMetaProps) {
    prevProps.meta.clear();

    this.renderMeta();
  }

  componentWillUnmount() {
    this.props.meta.clear();
  }

  private renderMeta() {
    const head = this.headRef?.current?.nextSibling;
    const tail = this.tailRef?.current;

    if (!head || !tail) {
      return;
    }

    this.props.meta.render(head, tail);
  }

  render() {
    return (
      <>
        {this.props.meta.length > 0 && <span style={{ display: 'none' }} ref={this.headRef} />}
        {this.props.children}
        {this.props.meta.length > 0 && <span style={{ display: 'none' }} ref={this.tailRef} />}
      </>
    );
  }
}
