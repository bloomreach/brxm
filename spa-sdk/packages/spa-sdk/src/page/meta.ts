/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const TYPE_META_COMMENT = 'comment';

/**
 * Meta-data following before a page component.
 */
export const META_POSITION_BEGIN = 'begin';

/**
 * Meta-data following after a page component.
 */
export const META_POSITION_END = 'end';

export type MetaType = typeof TYPE_META_COMMENT;
export type MetaPosition = typeof META_POSITION_BEGIN | typeof META_POSITION_END;

export interface MetaModel {
  data: string;
  type: MetaType;
}

/**
 * Meta information describing a part of the page.
 */
export interface Meta {
  /**
   * @return The meta-data.
   */
  getData(): string;

  /**
   * @return The meta-data position relative to the related element.
   */
  getPosition(): MetaPosition;
}

export class MetaImpl implements Meta {
  constructor(protected model: MetaModel, protected position: MetaPosition) {}

  getData() {
    return this.model.data;
  }

  getPosition() {
    return this.position;
  }
}

/**
 * Checks whether a value is a meta-data object.
 * @param value The value to check.
 */
export function isMeta(value: any): value is Meta {
  return value instanceof MetaImpl;
}
