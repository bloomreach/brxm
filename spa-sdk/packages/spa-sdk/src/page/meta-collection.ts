/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { inject, injectable } from 'inversify';
import { MetaFactory } from './meta-factory';
import { MetaModel, Meta, META_POSITION_BEGIN, META_POSITION_END } from './meta';
import { isMetaComment } from './meta-comment';

export const MetaCollectionModelToken = Symbol.for('MetaCollectionModelToken');

export interface MetaCollectionModel {
  beginNodeSpan?: MetaModel[];
  endNodeSpan?: MetaModel[];
}

/**
 * Collection of the meta-data describing a part of the page.
 * @note The collection extends the built-in Array type for backward compatibility.
 */
export interface MetaCollection extends Array<Meta> {
  /**
   * Clears all previously rendered meta-data objects.
   * @deprecated Use a callback returned by the `render` method.
   */
  clear(): void;

  /**
   * Renders meta-data objects on the page.
   * @param head The heading node of the page fragment.
   * @param tail The tailing node of the page fragment.
   * @return The callback to clear rendered meta-data objects.
   */
  render(head: Node, tail: Node): () => void;
}

@injectable()
export class MetaCollectionImpl extends Array<Meta> implements MetaCollection {
  private comments: Comment[] = [];

  constructor(
    @inject(MetaCollectionModelToken) model: MetaCollectionModel,
    @inject(MetaFactory) metaFactory: MetaFactory,
  ) {
    super(
      ...(model.beginNodeSpan || []).map(model => metaFactory.create(model, META_POSITION_BEGIN)),
      ...(model.endNodeSpan || []).map(model => metaFactory.create(model, META_POSITION_END)),
    );

    const prototype = Object.create(MetaCollectionImpl.prototype);

    prototype.constructor = Array.prototype.constructor;
    Object.setPrototypeOf(this, prototype);
    Object.freeze(this);
  }

  clear(comments = [...this.comments]) {
    comments.forEach((comment) => {
      comment.remove();

      const index = this.comments.indexOf(comment);
      if (index > -1) {
        this.comments.splice(index, 1);
      }
    });
  }

  render(head: Node, tail: Node) {
    const document = head.ownerDocument ?? tail.ownerDocument;
    const comments = document
      ? [
        ...this.filter(isMetaComment)
          .filter(meta => meta.getPosition() === META_POSITION_BEGIN)
          .map(meta => document.createComment(meta.getData()))
          .map((comment) => {
            head.parentNode?.insertBefore(comment, head);

            return comment;
          }),

        ...this.filter(isMetaComment)
          .filter(meta => meta.getPosition() === META_POSITION_END)
          .reverse()
          .map(meta => document.createComment(meta.getData()))
          .map((comment) => {
            if (tail.nextSibling) {
              tail.parentNode?.insertBefore(comment, tail.nextSibling);
            } else {
              tail.parentNode?.appendChild(comment);
            }

            return comment;
          }),
      ]
      : [];

    this.comments.push(...comments);

    return this.clear.bind(this, comments);
  }
}
