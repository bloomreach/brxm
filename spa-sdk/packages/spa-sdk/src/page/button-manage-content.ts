/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Content } from './content09';
import { Document } from './document';
import { Meta, MetaModel, META_POSITION_BEGIN, META_POSITION_END, TYPE_META_COMMENT } from './meta';
import { MetaCollectionModel, MetaCollection } from './meta-collection';

/**
 * A manage content button.
 */
export const TYPE_MANAGE_CONTENT_BUTTON = 'MANAGE_CONTENT_LINK';

export interface ManageContentButton {
  /**
   * The content entity to open for editing.
   */
  content?: Content | Document;

  /**
   * Template query to use for creating new documents.
   */
  documentTemplateQuery?: string;

  /**
   * Template query to use in case folders specified by `path` do not yet exist and must be created.
   */
  folderTemplateQuery?: string;

  /**
   * Initial location of a new document, relative to the `root`.
   */
  path?: string;

  /**
   * Name of the component parameter in which the document path is stored.
   */
  parameter?: string;

  /**
   * Flag indicating that the picked value should be stored as a relative path.
   */
  relative?: boolean;

  /**
   * Path to the root folder of selectable document locations.
   */
  root?: string;
}

export function createManageContentButton(params: ManageContentButton): MetaCollection | MetaCollectionModel {
  const meta = params.content?.getMeta();
  const entries = [
    ['defaultPath', params.path],
    ['documentTemplateQuery', params.documentTemplateQuery],
    ['folderTemplateQuery', params.folderTemplateQuery],
    ['rootPath', params.root],
    ['parameterName', params.parameter],
    ['parameterValueIsRelativePath', params.relative ? 'true' : undefined],
  ].filter(([, value]) => !!value);

  if (!entries.length) {
    return meta ?? {};
  }

  const model = Object.fromEntries(entries);

  if (!meta) {
    return {
      beginNodeSpan: [{
        type: TYPE_META_COMMENT,
        data: JSON.stringify({ 'HST-Type': TYPE_MANAGE_CONTENT_BUTTON, ...model }),
      }],
    };
  }

  const merge = (item: Meta) => ({
    type: TYPE_META_COMMENT,
    data: JSON.stringify(Object.assign(JSON.parse(item.getData()), model)),
  } as MetaModel);

  return {
    beginNodeSpan: meta.filter(item => item.getPosition() === META_POSITION_BEGIN).map(merge),
    endNodeSpan: meta.filter(item => item.getPosition() === META_POSITION_END).map(merge),
  };
}
