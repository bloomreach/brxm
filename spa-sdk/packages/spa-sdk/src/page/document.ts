/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { inject, injectable } from 'inversify';
import { LinkFactory } from './link-factory';
import { Link } from './link';
import { MetaCollectionFactory } from './meta-collection-factory';
import { MetaCollectionModel, MetaCollection } from './meta-collection';

export const DocumentModelToken = Symbol.for('DocumentModelToken');

export const TYPE_DOCUMENT = 'document';

type DocumentLinks = 'site';

interface DocumentDataModel {
  id: string;
  localeString?: string;
  name: string;
  [property: string]: any;
}

/**
 * Model of a document item.
 */
export interface DocumentModel {
  data: DocumentDataModel;
  links: Record<DocumentLinks, Link>;
  meta?: MetaCollectionModel;
  type: typeof TYPE_DOCUMENT;
}

/**
 * Document used on the page.
 */
export interface Document {
  /**
   * @return The document id.
   */
  getId(): string;

  /**
   * @return The document locale.
   */
  getLocale(): string | undefined;

  /**
   * @return The document meta-data collection.
   */
  getMeta(): MetaCollection;

  /**
   * @return The document name.
   */
  getName(): string;

  /**
   * @return The document data.
   */
  getData(): DocumentDataModel;
  getData<T extends Record<string, any>>(): T & DocumentDataModel;

  /**
   * @return The link to the content.
   */
  getUrl(): string | undefined;
}

@injectable()
export class DocumentImpl implements Document {
  protected meta: MetaCollection;

  constructor(
    @inject(DocumentModelToken) protected model: DocumentModel,
    @inject(LinkFactory) private linkFactory: LinkFactory,
    @inject(MetaCollectionFactory) metaFactory: MetaCollectionFactory,
  ) {
    this.meta = metaFactory(this.model.meta ?? {});
  }

  getId() {
    return this.model.data.id;
  }

  getLocale() {
    return this.model.data.localeString;
  }

  getMeta() {
    return this.meta;
  }

  getName() {
    return this.model.data.name;
  }

  getData() {
    return this.model.data;
  }

  getUrl() {
    return this.linkFactory.create(this.model.links.site);
  }
}

/**
 * Checks whether a value is a document.
 * @param value The value to check.
 */
export function isDocument(value: any): value is Document {
  return value instanceof DocumentImpl;
}
