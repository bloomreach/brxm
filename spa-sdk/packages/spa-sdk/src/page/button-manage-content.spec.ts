/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Document } from './document';
import { MetaCollection } from './meta-collection';
import { META_POSITION_BEGIN } from './meta';
import { TYPE_MANAGE_CONTENT_BUTTON, createManageContentButton } from './button-manage-content';

describe('createManageContentButton', () => {
  let content: jest.Mocked<Document>;
  let meta: jest.Mocked<MetaCollection>;

  beforeEach(() => {
    content = { getMeta: jest.fn(() => meta) } as unknown as typeof content;
    meta = [{
      getPosition: jest.fn(() => META_POSITION_BEGIN),
      getData: jest.fn(() => JSON.stringify({ 'HST-Type': TYPE_MANAGE_CONTENT_BUTTON, uuid: 'id' })),
    }] as unknown as typeof meta;
  });

  it('should return content meta-data as-is if there are no custom parameters', () => {
    expect(createManageContentButton({ content })).toBe(meta);
  });

  it('should return an empty meta-data model if there is no meta-data', () => {
    expect(createManageContentButton({})).toEqual({});
  });

  it('should generate add content button when there are no content parameters', () => {
    expect(createManageContentButton({
      documentTemplateQuery: 'new-content-document',
      folderTemplateQuery: 'new-content-folder',
      root: 'content',
    })).toMatchSnapshot();
  });

  it('should merge custom parameters', () => {
    expect(createManageContentButton({
      content,
      documentTemplateQuery: 'new-news-document',
      folderTemplateQuery: 'new-news-folder',
      root: 'news',
      path: '2020/11',
      parameter: 'document',
    })).toMatchSnapshot();
  });

  it('should set relative parameter flag', () => {
    expect(createManageContentButton({ content, relative: true })).toMatchSnapshot();
  });

  it('should not set relative parameter flag', () => {
    expect(createManageContentButton({ content, path: 'content', relative: false })).toMatchSnapshot();
    expect(createManageContentButton({ content, path: 'content' })).toMatchSnapshot();
  });
});
