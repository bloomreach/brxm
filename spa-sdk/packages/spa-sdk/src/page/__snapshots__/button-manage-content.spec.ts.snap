// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createManageContentButton should generate add content button when there are no content parameters 1`] = `
Object {
  "beginNodeSpan": Array [
    Object {
      "data": "{\\"HST-Type\\":\\"MANAGE_CONTENT_LINK\\",\\"documentTemplateQuery\\":\\"new-content-document\\",\\"folderTemplateQuery\\":\\"new-content-folder\\",\\"rootPath\\":\\"content\\"}",
      "type": "comment",
    },
  ],
}
`;

exports[`createManageContentButton should merge custom parameters 1`] = `
Object {
  "beginNodeSpan": Array [
    Object {
      "data": "{\\"HST-Type\\":\\"MANAGE_CONTENT_LINK\\",\\"uuid\\":\\"id\\",\\"defaultPath\\":\\"2020/11\\",\\"documentTemplateQuery\\":\\"new-news-document\\",\\"folderTemplateQuery\\":\\"new-news-folder\\",\\"rootPath\\":\\"news\\",\\"parameterName\\":\\"document\\"}",
      "type": "comment",
    },
  ],
  "endNodeSpan": Array [],
}
`;

exports[`createManageContentButton should not set relative parameter flag 1`] = `
Object {
  "beginNodeSpan": Array [
    Object {
      "data": "{\\"HST-Type\\":\\"MANAGE_CONTENT_LINK\\",\\"uuid\\":\\"id\\",\\"defaultPath\\":\\"content\\"}",
      "type": "comment",
    },
  ],
  "endNodeSpan": Array [],
}
`;

exports[`createManageContentButton should not set relative parameter flag 2`] = `
Object {
  "beginNodeSpan": Array [
    Object {
      "data": "{\\"HST-Type\\":\\"MANAGE_CONTENT_LINK\\",\\"uuid\\":\\"id\\",\\"defaultPath\\":\\"content\\"}",
      "type": "comment",
    },
  ],
  "endNodeSpan": Array [],
}
`;

exports[`createManageContentButton should set relative parameter flag 1`] = `
Object {
  "beginNodeSpan": Array [
    Object {
      "data": "{\\"HST-Type\\":\\"MANAGE_CONTENT_LINK\\",\\"uuid\\":\\"id\\",\\"parameterValueIsRelativePath\\":\\"true\\"}",
      "type": "comment",
    },
  ],
  "endNodeSpan": Array [],
}
`;
